server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateMetaData.java

```diff
@@ -331,19 +330,40 @@ public IndexTemplateMetaData build() {
             return new IndexTemplateMetaData(name, order, version, indexPatterns, settings, mappings.build(), aliases.build());
         }
 
-        public static void toXContent(IndexTemplateMetaData indexTemplateMetaData, XContentBuilder builder, ToXContent.Params params)
-                throws IOException {
+        public static void toXContent(IndexTemplateMetaData indexTemplateMetaData,
+                                      XContentBuilder builder,
+                                      ToXContent.Params params) throws IOException {
             builder.startObject(indexTemplateMetaData.name());
+            toInnerXContent(indexTemplateMetaData, builder, params, true);
+            builder.endObject();
+        }
 
-            toInnerXContent(indexTemplateMetaData, builder, params);
-
+        /**
+         * Serializes the template to xContent, making sure not to nest mappings under the
+         * type name.
+         *
+         * Note that this method should currently only be used for creating REST responses,
+         * and not when directly updating stored templates. Index templates are still stored
+         * in the old, typed format, and have yet to be migrated to be typeless.
+         */
+        public static void toXContentWithoutTypes(IndexTemplateMetaData indexTemplateMetaData,
+                                                  XContentBuilder builder,
+                                                  ToXContent.Params params) throws IOException {
+            builder.startObject(indexTemplateMetaData.name());
+            toInnerXContent(indexTemplateMetaData, builder, params, false);
             builder.endObject();
         }
 
-        public static void toInnerXContent(IndexTemplateMetaData indexTemplateMetaData, XContentBuilder builder, ToXContent.Params params)
-            throws IOException {
+        public static void toInnerXContent(IndexTemplateMetaData indexTemplateMetaData,
+                                           XContentBuilder builder,
+                                           ToXContent.Params params) throws IOException {
+            toInnerXContent(indexTemplateMetaData, builder, params, true);
```

Comment:
Curious to understand under whoch circumstance we use "includeTypeName" true here. Seems to come from TemplateUpgradeService. Can the method be package private to avoid any other accidental internal use?

Reply:
Yes, because `TemplateUpgradeService` works with the actual templates as they are stored, it needs to read and write 'typed' mappings. I will make this package-private as you suggested, and likely add some more Javadoc to clarify.